File: frontend/src/api/finance.ts
import apiClient from './client';

export interface FinanceSummary {
  date_from: string;
  date_to: string;
  total_revenue: number;
  total_cogs: number;
  gross_profit: number;
  gross_margin_pct: number;
  total_wastage_cost: number;
  estimated_net_profit: number;
  total_units_sold: number;
  total_wastage_units: number;
  missing_cost_rows: number;
}

export interface ProductMarginItem {
  product_id: string;
  product_name: string;
  sku: string;
  units_sold: number;
  revenue: number;
  cogs: number;
  gross_profit: number;
  gross_margin_pct: number;
  avg_selling_price: number;
  unit_cogs: number;
  missing_cost: boolean;
}

export interface ProductMarginResponse {
  date_from: string;
  date_to: string;
  items: ProductMarginItem[];
}

export interface PnlTrendPoint {
  date: string;
  revenue: number;
  cogs: number;
  gross_profit: number;
  wastage_cost: number;
  estimated_net_profit: number;
}

export interface PnlTrendResponse {
  date_from: string;
  date_to: string;
  points: PnlTrendPoint[];
  total_revenue: number;
  total_cogs: number;
  gross_profit: number;
  total_wastage_cost: number;
  estimated_net_profit: number;
}

export interface FinanceFilters {
  date_from?: string;
  date_to?: string;
  store_id?: string;
  product_id?: string;
}

export const financeApi = {
  getSummary: async (params?: FinanceFilters): Promise<FinanceSummary> => {
    const response = await apiClient.get('/finance/summary', { params });
    return response.data;
  },

  getProductMargins: async (
    params?: Omit<FinanceFilters, 'product_id'> & { limit?: number }
  ): Promise<ProductMarginResponse> => {
    const response = await apiClient.get('/finance/product-margins', { params });
    return response.data;
  },

  getPnlTrend: async (params?: FinanceFilters): Promise<PnlTrendResponse> => {
    const response = await apiClient.get('/finance/pnl-trend', { params });
    return response.data;
  },
};

File: backend/app/schemas/finance.py
from datetime import date
from typing import List
from uuid import UUID

from pydantic import BaseModel


class FinanceSummaryResponse(BaseModel):
    date_from: date
    date_to: date
    total_revenue: float
    total_cogs: float
    gross_profit: float
    gross_margin_pct: float
    store_wastage_cost: float
    ingredient_wastage_cost: float
    production_product_wastage_cost: float
    production_wastage_cost: float
    total_wastage_cost: float
    estimated_net_profit: float
    total_units_sold: int
    total_wastage_units: int
    missing_cost_rows: int
    estimated_cost_rows: int


class ProductMarginItem(BaseModel):
    product_id: UUID
    product_name: str
    sku: str
    units_sold: int
    revenue: float
    cogs: float
    gross_profit: float
    gross_margin_pct: float
    avg_selling_price: float
    unit_cogs: float
    missing_cost: bool
    estimated_cost: bool


class ProductMarginResponse(BaseModel):
    date_from: date
    date_to: date
    items: List[ProductMarginItem]


class PnlTrendPoint(BaseModel):
    date: date
    revenue: float
    cogs: float
    gross_profit: float
    store_wastage_cost: float
    ingredient_wastage_cost: float
    production_product_wastage_cost: float
    production_wastage_cost: float
    wastage_cost: float
    estimated_net_profit: float


class PnlTrendResponse(BaseModel):
    date_from: date
    date_to: date
    points: List[PnlTrendPoint]
    total_revenue: float
    total_cogs: float
    gross_profit: float
    total_store_wastage_cost: float
    total_ingredient_wastage_cost: float
    total_production_product_wastage_cost: float
    total_production_wastage_cost: float
    total_wastage_cost: float
    estimated_net_profit: float
